Remove hdf4 support from NexusCpp

peterfpeterson opened the pull request:
This removes hdf4 from `FrameworkNexus` in its entirety. 

LoadILLTOF2 was converted to use LegacyNexus (to support legacy hdf4 files) and LoadILLTOF3 is a copy that uses Nexus (hdf5). There is also a new set of tests for `FileLoaderRegistry` which helped figure out which loaders support which backend.

The change to `LegacyNexus`'s make is because I discovered it was using [`set_target_properties`](https://cmake.org/cmake/help/latest/command/set_target_properties.html) incorrectly and the flag was making it in through find cmake find modules.

Refs #38332

### To test:

This is mostly a refactor of code and additional unit tests were added to help verify it is working. 

*This does not require release notes* because it is part of the nexus API consolidation which will have a more extensive write-up.

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

peterfpeterson committed:
Remove hdf4 support from NexusCpp

rosswhitfield committed:
Fix napi.h in LegacyNexus namespace

rosswhitfield committed:
Update or remove test files

rosswhitfield committed:
Move LoadNexus to use LegacyNexus

peterfpeterson committed:
Move to LegacyNexusDescrpitor because some IN5 files are hdf4

peterfpeterson committed:
Add broken test for hdf4 files

perenon committed:
Adding LoadILLTOF3 based on HDF5

peterfpeterson committed:
Move LegacyNexusClasses into LegacyNexus

peterfpeterson committed:
Get LoadILLTOF2 working with LegacyNexus

This required copying a bunch of code from LoadHelper over into the
LoadILLTOF2 cpp file.

peterfpeterson committed:
Minor code cleanup

rboston628 commented on Framework/DataHandling/src/LoadNexus.cpp:
```diff
@@ -293,7 +297,7 @@ int LoadNexus::getNexusEntryTypes(const std::string &fileName, std::vector<std::
   while (true) {
     entry = fileH->getNextEntry();
-    if (entry == ::NeXus::EOD_ENTRY)
+    if (entry.first == NULL_STR && entry.second == NULL_STR)
```
I really prefer the old checks, but I guess that isn't available in LegacyNexus.  It wouldn't be hard to add ,though

rboston628 commented on Framework/DataHandling/test/LoadDetectorInfoTest.h:
```diff
@@ -262,7 +262,7 @@ class LoadDetectorInfoTest : public CxxTest::TestSuite {
     makeTestWorkspace(SmallTestDatFile::NDETECTS, NBINS, m_InoutWS);
     loadDetInfo.setPropertyValue("Workspace", m_InoutWS);
-    loadDetInfo.setPropertyValue("DataFilename", "argus0026287.nxs");
+    loadDetInfo.setPropertyValue("DataFilename", "ARGUS00073601.nxs");
```
Why was this change necessary?

peterfpeterson replied:
That was a fix from Ross. I assume that it was an issue with `FileFinder` / `ArchiveSearch` forgetting what the instrument is

rboston628 replied:
But doesn't changing the file used for the test, effectively change what the test is testing?  If it was necessary, it suggests some behavior changed

peterfpeterson replied:
I'm not sure why. It is a Ross change

rboston628 replied:
@rosswhitfield  Was this change necessary, or is this file just faster to work with?

rboston628 replied:
Doing manual testing, it seems the issue is that `argus0026287.nxs` can no longer be loaded by `LoadILLTOF2`.  It generates errors within `File::openData()`.  This might be due to a different HDF version.  Is this chagne in behavior anticipated and properly accounted for?

rboston628 replied:
@MialLewis  is this instrument a muon instrument?

peterfpeterson replied:
This appears to be another hidden hdf4 based usage. For most everything else we've copied the code into the LegacyNexus portion of things. However, the way `LoadDetectorInfo` is used by muons isn't entirely clear. We need @MialLewis's advice
```sh
$  rg LoadDetectorInfo -l -t cpp -t py | grep -v LoadDetectorInfo
scripts/Inelastic/Direct/RunDescriptor.py
Testing/SystemTests/tests/framework/LoadLotsOfFiles.py
Framework/WorkflowAlgorithms/src/DgsConvertToEnergyTransfer.cpp
```

rosswhitfield replied:
Yeah I changed the test data from a HDF4 file to HDF5 just to get the test to pass but we should check if they need `LoadDetectorInfo` to support HDF4.

I also removed the HDF4 file from the `StartAndEndTimeFromNexusFileExtractorTest` because I didn't see any HDF4 algorithms using the `extractStartTime` or `extractEndTime` functions.

peterfpeterson replied:
Documenting the discussion that happened on slack. This particular issue (ARGUS) is being resolved by deleting it for now. If we need it later, the code will be resurrected.

rboston628 commented on Framework/DataHandling/test/LoadILLTOF2Test.h:
```diff
@@ -82,9 +82,11 @@ class LoadILLTOF2Test : public CxxTest::TestSuite {
       TS_ASSERT_EQUALS(histogram.size(), numberOfChannels)
       const auto &xs = histogram.x();
       if (convertToTOF) {
+        const auto mywidth = xs[1] - xs[0];
+        const auto myoffset = xs[0] - (mywidth / 2.);
```
I don't see where these variables are being used

peterfpeterson replied:
Oops! Those were for debugging the right parameter values. I reverse engineered the answers. A compiler should have given me warnings about unused variables.

rboston628 replied:
Yeah, I'm not sure how that got past both compiler and cppcheck

rboston628 commented on Framework/DataHandling/test/StartAndEndTimeFromNexusFileExtractorTest.h:
```diff
@@ -32,14 +32,6 @@ class StartAndEndTimeFromNexusFileExtractorTest : public CxxTest::TestSuite {
                                extractEndTime(filename));
   }
 
-  void test_that_times_can_be_extracted_from_muon_file() {
-    const std::string filename("emu00006473.nxs");
-    const std::string startTime = "2006-11-21T07:04:30";
-    const std::string endTime = "2006-11-21T09:29:28";
-
-    do_test(filename, startTime, endTime);
-  }
```
Why was this  test removed?

peterfpeterson replied:
EMU is a muon instrument with an HDF4 backed nexus file. It isn't using this function anymore because it is handled by `Framework/Muon` now.

rboston628 commented on Framework/API/test/FileLoaderRegistryTest.h:
```diff
@@ -0,0 +1,72 @@
+
+    TS_ASSERT(!path.empty());
+
+    // algorithm can load the file
+    // TS_ASSERT(FileLoaderRegistry::Instance().canLoad(alg_expected, path));
```
Should this be commented?

rboston628 commented on Framework/DataHandling/src/LoadILLTOF2.cpp:
```diff
@@ -15,23 +15,304 @@
+  std::vector<NXClassInfo> v = firstEntry.groups();
+  const auto it = std::find_if(v.cbegin(), v.cend(), [](const auto &group) { return group.nxclass == NXINSTRUMENT; });
+  if (it != v.cend())
+    return it->nxname;
+  return "";
```
Fix for single return

rboston628 commented on Framework/DataHandling/src/LoadILLTOF2.cpp:
I'm not sure why all of these new methods are needed.  It doesn't look like they get called.

rboston628 reviewed (commented):
I'm not sure how to review LoadILLTOF3, and will need to do that tomorrow.  Here are some comments on the rest

peterfpeterson committed:
Copy over EOD_ENTRY from FrameworkNexus for clearer code

peterfpeterson committed:
Modify assert to give clear error message

peterfpeterson committed:
Minor code cleanup

peterfpeterson committed:
Register the loader correctly

peterfpeterson committed:
Remove HDF4 from the nexus tests again

peterfpeterson committed:
Do not suggest a loader

By not suggesting a loader, this uses generic load which correctly
selects the correct version of LoadILLTOF.

rboston628 reviewed (approved):
I am accepting the ARGUS issue as resolved.  That was my only hesitation.  All looks good.
